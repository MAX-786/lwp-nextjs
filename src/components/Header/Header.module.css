
.header {
  background-color: #ffffff;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  padding: 1rem 0;
  position: sticky;
  top: 0;
  z-index: 100;
}

.nav {
  max-width: 1200px;
  margin: 0 auto;
  padding: 0 1rem;
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.navList {
  display: flex;
  gap: 2rem;
  list-style: none;
  margin: 0;
  padding: 0;
}

.navItem {
  display: flex;
  align-items: center;
}

.navLink {
  color: #333;
  text-decoration: none;
  font-weight: 500;
  transition: color 0.2s;
}

.navLink:hover {
  color: #0070f3;
}

.logoutButton {
  background: none;
  border: none;
  color: #333;
  font: inherit;
  cursor: pointer;
  padding: 0;
}

.logoutButton:hover {
  color: #0070f3;
}

@media (max-width: 768px) {
  .nav {
    flex-direction: column;
    gap: 1rem;
  }
  
  .navList {
    gap: 1rem;
    flex-wrap: wrap;
    justify-content: center;
  }
}