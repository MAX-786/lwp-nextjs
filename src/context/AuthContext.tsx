"use client";
import React, { createContext, useContext, useState } from 'react';
import axios from 'axios';

interface AuthContextType {
  user: string | null;
  login: (email: string, password: string) => void;
  register: (email: string, password: string) => void;
  logout: () => void;
}

const AuthContext = createContext<AuthContextType | undefined>(undefined);

export const AuthProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [user, setUser] = useState<string | null>(null);

  const login = async (email: string, password: string) => {
    try {
      const response = await axios.post('/api/login', { email, password });
      setUser(response.data.email);
    } catch (error) {
      console.error('Login failed:', error);
    }
  };

  const register = async (email: string, password: string) => {
    try {
      const response = await axios.post('/api/register', { email, password });
      setUser(response.data.email);
    } catch (error) {
      console.error('Registration failed:', error);
    }
  };

  const logout = () => {
    setUser(null);
  };

  return (
    <AuthContext.Provider value={{ user, login, register, logout }}>
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (!context) {
    throw new Error('useAuth must be used within an AuthProvider');
  }
  return context;
};