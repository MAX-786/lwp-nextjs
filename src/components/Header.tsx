"use client";
import React from "react";
import Link from "next/link";
import { useAuth } from "@/context/AuthContext";
import { useRouter } from "next/navigation"; // Use the navigation router from "next/navigation".

const Header: React.FC = () => {
  const { user, logout } = useAuth();
  const isLoggedIn = user !== null;

  // Ensure useRouter is accessed only in client-side rendering
  let currentPath = "/";
  if (typeof window !== "undefined") {
    const router = useRouter();
    currentPath = router.pathname; // Get the current route
  }

  return (
    <header>
      <nav>
        <ul>
          <li className={currentPath === "/" ? "active" : ""}>
            <Link href="/">Home</Link>
          </li>
          {isLoggedIn ? (
            <li>
              <button onClick={logout}>Logout</button>
            </li>
          ) : (
            <>
              <li className={currentPath === "/login" ? "active" : ""}>
                <Link href="/login">Login</Link>
              </li>
              <li className={currentPath === "/register" ? "active" : ""}>
                <Link href="/register">Register</Link>
              </li>
            </>
          )}
          <li className={currentPath === "/todos" ? "active" : ""}>
            <Link href="/todos">Todos</Link>
          </li>
        </ul>
      </nav>

      {/* Add CSS for styling */}
      <style jsx>{`
        header {
          padding: 1rem;
          background: #f5f5f5;
        }
        nav ul {
          display: flex;
          list-style: none;
          padding: 0;
          margin: 0;
        }
        nav li {
          margin-right: 15px;
        }
        nav a {
          text-decoration: none;
          color: black;
        }
        .active a {
          font-weight: bold;
          color: blue;
        }
      `}</style>
    </header>
  );
};

export default Header;
