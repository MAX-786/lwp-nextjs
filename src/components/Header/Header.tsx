"use client";
import React from "react";
import Link from "next/link";
import { useAuth } from "@/context/AuthContext";
import styles from "./Header.module.css";

const Header: React.FC<object> = () => {
  const { user, logout } = useAuth();
  const isLoggedIn = user !== null;
  
  return (
    <header className={styles.header}>
      <nav className={styles.nav}>
        <ul className={styles.navList}>
          <li className={styles.navItem}>
            <Link href="/" className={styles.navLink}>Home</Link>
          </li>
          <li className={styles.navItem}>
            <Link href="/todos" className={styles.navLink}>Todos</Link>
          </li>
          {isLoggedIn ? (
            <li className={styles.navItem}>
              <button onClick={logout} className={styles.logoutButton}>Logout</button>
            </li>
          ) : (
            <>
              <li className={styles.navItem}>
                <Link href="/login" className={styles.navLink}>Login</Link>
              </li>
              <li className={styles.navItem}>
                <Link href="/register" className={styles.navLink}>Register</Link>
              </li>
              <li className={styles.navItem}>
                <Link href="/chats" className={styles.navLink}>Chats</Link>
              </li>
            </>
          )}
        </ul>
      </nav>
    </header>
  );
};

export default Header;