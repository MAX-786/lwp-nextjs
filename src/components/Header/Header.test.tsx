// Header.test.tsx
import { render, screen, fireEvent } from "@testing-library/react";
import Header from "./Header";
import { useAuth } from "@/context/AuthContext";

// Mock the useAuth hook
jest.mock("@/context/AuthContext", () => ({
  useAuth: jest.fn(() => ({
    user: null,
    logout: jest.fn(),
  })),
}));

// Mock Next.js Link component
jest.mock("next/link", () => {
  // eslint-disable-next-line react/display-name
  return ({ children, href }: { children: React.ReactNode; href: string }) => {
    return <a href={href}>{children}</a>;
  };
});

// Mock CSS modules
jest.mock("./Header.module.css", () => ({
  header: "header",
  nav: "nav",
  navList: "navList",
  navItem: "navItem",
  navLink: "navLink",
  logoutButton: "logoutButton",
}));

describe("Header Component", () => {
  const mockLogout = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe("When user is logged out", () => {
    beforeEach(() => {
      (useAuth as jest.Mock).mockReturnValue({
        user: null,
        logout: mockLogout,
      });
      render(<Header />);
    });

    it("renders the basic navigation links", () => {
      expect(screen.getByRole("link", { name: "Home" })).toHaveAttribute(
        "href",
        "/"
      );
      expect(screen.getByRole("link", { name: "Todos" })).toHaveAttribute(
        "href",
        "/todos"
      );
      expect(screen.getByRole("link", { name: "Chats" })).toHaveAttribute(
        "href",
        "/chats"
      );
    });

    it("shows login and register links", () => {
      expect(screen.getByRole("link", { name: "Login" })).toHaveAttribute(
        "href",
        "/login"
      );
      expect(screen.getByRole("link", { name: "Register" })).toHaveAttribute(
        "href",
        "/register"
      );
    });

    it("does not show logout button", () => {
      expect(
        screen.queryByRole("button", { name: "Logout" })
      ).not.toBeInTheDocument();
    });
  });

  describe("When user is logged in", () => {
    beforeEach(() => {
      (useAuth as jest.Mock).mockReturnValue({
        user: { id: "123", email: "test@example.com" },
        logout: mockLogout,
      });
      render(<Header />);
    });

    it("shows logout button", () => {
      expect(
        screen.getByRole("button", { name: "Logout" })
      ).toBeInTheDocument();
    });

    it("hides login and register links", () => {
      expect(
        screen.queryByRole("link", { name: "Login" })
      ).not.toBeInTheDocument();
      expect(
        screen.queryByRole("link", { name: "Register" })
      ).not.toBeInTheDocument();
      expect(
        screen.queryByRole("link", { name: "Chats" })
      ).toBeInTheDocument();
    });

    it("triggers logout when clicking logout button", () => {
      const logoutButton = screen.getByRole("button", { name: "Logout" });
      fireEvent.click(logoutButton);
      expect(mockLogout).toHaveBeenCalledTimes(1);
    });
  });

  it("matches snapshot when logged out", () => {
    const { asFragment } = render(<Header />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("matches snapshot when logged in", () => {
    (useAuth as jest.Mock).mockReturnValue({
      user: { id: "123", email: "test@example.com" },
      logout: mockLogout,
    });
    const { asFragment } = render(<Header />);
    expect(asFragment()).toMatchSnapshot();
  });
});
