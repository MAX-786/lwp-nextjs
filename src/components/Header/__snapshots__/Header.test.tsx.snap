// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component matches snapshot when logged in 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <nav
      class="nav"
    >
      <ul
        class="navList"
      >
        <li
          class="navItem"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="navItem"
        >
          <a
            href="/todos"
          >
            Todos
          </a>
        </li>
        <li
          class="navItem"
        >
          <a
            href="/chats"
          >
            Chats
          </a>
        </li>
        <li
          class="navItem"
        >
          <button
            class="logoutButton"
          >
            Logout
          </button>
        </li>
      </ul>
    </nav>
  </header>
</DocumentFragment>
`;

exports[`Header Component matches snapshot when logged out 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <nav
      class="nav"
    >
      <ul
        class="navList"
      >
        <li
          class="navItem"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="navItem"
        >
          <a
            href="/todos"
          >
            Todos
          </a>
        </li>
        <li
          class="navItem"
        >
          <a
            href="/chats"
          >
            Chats
          </a>
        </li>
        <li
          class="navItem"
        >
          <button
            class="logoutButton"
          >
            Logout
          </button>
        </li>
      </ul>
    </nav>
  </header>
</DocumentFragment>
`;
